openvdb/util/CpuTimer.h

```diff
@@ -56,65 +57,130 @@ namespace util {
 ///
 /// @code
-///    CpuTimer timer("algorithm 1");
+///    util::CpuTimer timer("algorithm 1");
 ///    // code to be timed goes here
 ///    timer.restart("algorithm 2");
 ///    // code to be timed goes here
 ///    timer.stop();
 /// @endcode
+///
+/// or to measure speedup between multiple runs
+///
+/// @code
+///    util::CpuTimer timer("algorithm 1");
+///    // code for the first run goes here
+///    const double t1 = timer.restart("algorithm 2");
+///    // code for the second run goes here
+///    const double t2 = timer.stop();
+///    std::cerr << "Algorithm 1 is " << (t2/t1)
+///              << " timers faster than algorithm 2\n";
+/// @endcode
+///
+/// or to measure multiple blocks of code with deferred output
+///
+/// @code
+///    util::CpuTimer timer();
+///    // code here will not be timed!
+///    timer.start();
+///    // code for the first run goes here
+///    const double t1 = timer.restart();//time in milliseconds
+///    // code for the second run goes here
+///    const double t2 = timer.restart();//time in milliseconds
+///    // code here will not be timed!
+///    util::printTime(std::cout, t1, "Algorithm 1 completed in ");
+///    util::printTime(std::cout, t2, "Algorithm 2 completed in ");
+/// @endcode
 class CpuTimer
 {
 public:
 
     /// @brief Initiate timer
-    CpuTimer() : mT0(tbb::tick_count::now()) {}
+    CpuTimer(std::ostream& os = std::cerr) : mOutStream(os), mT0(tbb::tick_count::now()) {}
 
-    /// @brief Prints message and re-start timer.
+    /// @brief Prints message and start timer.
     ///
     /// @note Should normally be followed by a call to stop()
-    CpuTimer(const std::string& msg) { this->start(msg); }
+    CpuTimer(const std::string& msg, std::ostream& os = std::cerr) : mOutStream(os), mT0() { this->start(msg); }
 
     /// @brief Start timer.
     ///
-    /// @note Should normally be followed by a call to time()
+    /// @note Should normally be followed by a call to milliseconds() or stop(std::string)
     inline void start() { mT0 = tbb::tick_count::now(); }
 
-    /// @brief Print message and re-start timer.
+    /// @brief Print message and start timer.
     ///
     /// @note Should normally be followed by a call to stop()
     inline void start(const std::string& msg)
     {
-        std::cerr << msg << " ... ";
+        mOutStream << msg << " ...";
         this->start();
     }
 
-    /// @brief Stop previous timer, print message and re-start timer.
+    /// @brief Return Time difference in milliseconds since construction or start was called.
     ///
-    /// @note Should normally be followed by a call to stop()
-    inline void restart(const std::string& msg)
+    /// @note Combine this method with start() to get timing without any outputs.
+    inline double milliseconds() const
     {
-        this->stop();
-        this->start(msg);
+        tbb::tick_count::interval_t dt = tbb::tick_count::now() - mT0;
+        return 1000.0*dt.seconds();
     }
```

Comment:
Any chance we could add a seconds() method to this class too? I'm using this in a few places where the time scale is much larger so end up having to multiple delta() by 1000 everywhere. Could go even further and introduce something like an enum which defines the return scale:

```
inline double detla(const Precision p = MILLISECONDES) {
  tbb::tick_count::interval_t dt = tbb::tick_count::now() - mT0;
  double s = dt.seconds();
  switch (p) {
    case MILLISECONDES : return s*1000.0;
    // etc
   }
}
```

But if not, something which just returns dt.seconds() would be great

Reply:
added

Reply:
I'm deprecating the delta() method in favor of a seconds() and miliseconds() method